File: python-backend/rag_uploadfiles.py
from fastapi import APIRouter, UploadFile, File
from fastapi.responses import JSONResponse
from langchain_text_splitters import RecursiveCharacterTextSplitter
import io
from langchain_community.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
import os
import uuid
import tempfile
from dotenv import load_dotenv
from opensearchpy import OpenSearch
from opensearchpy.helpers import bulk
from openai import OpenAI
from util import generateToken

generateToken()

router = APIRouter()

host = os.environ.get("OPENSEARCH_HOST")
port = os.environ.get("OPENSEARCH_PORT")
username = os.environ.get("OPENSEARCH_USERNAME")
password = os.environ.get("OPENSEARCH_PASSWORD")

# OpenSearch configuration
OPENSEARCH_CONFIG = {
    "hosts": [{"host": host, "port": port}],
    "http_auth": (username, password),
    "http_compress": True,
    "use_ssl": True,
    "verify_certs": False,
    "ssl_assert_hostname": False,
    "ssl_show_warn": False,
}

INDEX_NAME = "files"

def generate_embeddings(texts):
    client = OpenAI()  # Get the OpenAI client
    # Generate embeddings using OpenAI API
    response = client.embeddings.create(
        input=texts, dimensions=256, model="text-embedding-3-large"
    )
    # Extract embeddings from the response
    embeddings = [item.embedding for item in response.data]
    return embeddings


def insert_documents(search_client, fileChunks, embeddings, fileMetadata):
    documents = []  # List to hold bulk actions
    for i, item in enumerate(fileMetadata):
        # Create an document object   
        document = {
            "_index": INDEX_NAME,
            "_id": str(uuid.uuid4()),  # Generate a unique GUID for the document ID
            "_source": {
                "name":  os.path.basename(item["source"]),  # Document source name without absolute path
                "content": fileChunks[i],  # Document content
                "embedding": embeddings[i],  # Document embedding
            },
        }
        documents.append(document)  # Add document to the list

    # Perform bulk insert into OpenSearch
    success, _ = bulk(search_client, documents)
    print(f"Successfully inserted {success} documents into OpenSearch.")


#Challenge 1: Implement the retrieve_all_documents function to retrieve all documents from OpenSearch index files.
def retrieve_all_documents(client):
    # Perform the OpenSearch search to get ALL documents
    # query should as here: "query": {"match_all": {}},
    search_body = {
        
    }
    response = client.search(index=INDEX_NAME, body=search_body)
    return response


@router.get("/rag/files")
async def get_files():
    try:
        # Connect to OpenSearch
        client = OpenSearch(**OPENSEARCH_CONFIG)
        search_results = retrieve_all_documents(client)
        # Challenge2: Get hits (replace "" with relevant code) Note: Debug and find the object structure from search results.
        hits = ""
        # To get only distinct files
        sources = []  # List to hold file names
        for hit in hits:
            source_name = hit["_source"]["name"]
            fileName = os.path.basename(source_name)
            if fileName not in sources:
                sources.append(fileName)  # To Remove the absolute path and get only the file name
        return JSONResponse(content={"sources": sources})
    except Exception as e:
        return JSONResponse(content={"error": str(e)}, status_code=500)

@router.post("/rag/upload")
 # Save the text content to OpenSearch index by name files
async def upload_file(file: UploadFile = File(...)):
    if not file.filename.endswith(".pdf"):
        return JSONResponse(
            content={"error": "Only PDF files are allowed"}, status_code=400
        )
    try:
        # Read the uploaded file
        contents = await file.read()
        pdf_stream = io.BytesIO(contents)

        # Get original filename and sanitize it
        filenamewithoutExt = file.filename.rsplit(".", 1)[0]  # Remove the file extension
        safe_filename = "".join(
            [c for c in filenamewithoutExt if c.isalnum() or c in ("-", "_")]
        )
        safe_filename += ".pdf"  # Ensure the filename ends with .pdf

        # Create temporary file with original name in temp directory
        temp_path = os.path.join(tempfile.gettempdir(), safe_filename)
        # Write the PDF content to the temporary file
        with open(temp_path, "wb") as temp_file:
            temp_file.write(pdf_stream.getvalue())

        try:
            # Use temporary file with PyPDFLoader
            loader = PyPDFLoader(temp_path)
            text_content = loader.load()

            # Split the docs and create embeddings
            text_splitter = RecursiveCharacterTextSplitter(
                chunk_size=3000, chunk_overlap=100
            )
            # Split the text content into chunks
            chunks = text_splitter.split_documents(text_content)

            pages_content = []
            pages_metadata_content = []
            for chunk in chunks:
                pages_content.append(chunk.page_content)
                pages_metadata_content.append(chunk.metadata)

            # Generate embeddings for the pages
            embeddings = generate_embeddings(pages_content)
            # Connect to OpenSearch
            client = OpenSearch(**OPENSEARCH_CONFIG)
            # Insert the documents into OpenSearch
            insert_documents(client, pages_content, embeddings, pages_metadata_content)
        finally:
            # Clean up temporary file
            os.unlink(temp_path)
        return JSONResponse(
            content={
                "message": "File uploaded and processed successfully",
                "filename": file.filename,
                "size": len(contents),
                "text_length": len(text_content),
            }
        )
    except Exception as e:
        return JSONResponse(content={"error": str(e)}, status_code=500)

File: nodejs-backend/routes/rag_uploadfiles.js
import express from 'express';
import multer from 'multer';
import fs from 'fs';
import path from 'path';
import os from 'os';
import { Readable } from 'stream';
import { Client as OpenSearchClient } from '@opensearch-project/opensearch';
import { generateToken } from '../util.js';
import { PDFLoader } from "@langchain/community/document_loaders/fs/pdf";
import { v4 as uuidv4 } from 'uuid';
import dotenv from "dotenv";

dotenv.config();
const router = express.Router();
await generateToken();

const upload = multer();
const INDEX_NAME = 'files';

const OPENSEARCH_CONFIG = {
  node: `https://${process.env.OPENSEARCH_HOST}:${process.env.OPENSEARCH_PORT}`,
  auth: {
    username: process.env.OPENSEARCH_USERNAME,
    password: process.env.OPENSEARCH_PASSWORD,
  },
  ssl: {
    rejectUnauthorized: false,
  },
};

export async function generate_embeddings(texts) {
  const client = await OpenAI();
  const response = await client.embeddings.create({
    input: texts,
    model: 'text-embedding-3-large',
    dimensions: 256
  });
  return response.data.map(item => item.embedding);
}

async function insert_documents(search_client, fileChunks, embeddings, fileMetadata) {
  const documents = fileChunks.map((content, i) => ({
    index: {
      _index: INDEX_NAME,
      _id: uuidv4(),
      _source: {
        name: path.basename(fileMetadata[i].source),
        content,
        embedding: embeddings[i],
      },
    },
  }));
  const { body: success } = await search_client.helpers.bulk({
    index: INDEX_NAME, // the default index
    datasource: documents,
    onDocument (doc) {
      return {
        index: { _index: INDEX_NAME }
      }
    }
  });
  console.log(`Successfully inserted ${success.items.length} documents into OpenSearch.`);
}

//Challenge 1: Implement the retrieve_all_documents function to retrieve all documents from OpenSearch index files.
async function retrieve_all_documents(client) {
  const search_body = {
    
  };
  const { body: response } = await client.search({ index: INDEX_NAME, body: search_body });
  return response;
}

router.get('/rag/files', async (req, res) => {
  try {
    const client = new OpenSearchClient({
     ...OPENSEARCH_CONFIG,
    });

    const search_results = await retrieve_all_documents(client);
    //Challenge2: Get hits (replace "" with relevant code) Note: Debug and find the object structure from search results.
    const sources = [];
    res.json({ sources });
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

router.post('/rag/upload', upload.single('file'), async (req, res) => {
  const file = req.file;
  if (!file || !file.originalname.endsWith('.pdf')) {
    return res.status(400).json({ error: 'Only PDF files are allowed' });
  }

  try {
    const pdf_stream = new Readable();
    pdf_stream.push(file.buffer);
    pdf_stream.push(null);

    const filenamewithoutExt = path.basename(file.originalname, path.extname(file.originalname));
    const safe_filename = filenamewithoutExt.replace(/[^a-z0-9-_]/gi, '_') + '.pdf';
    const temp_path = path.join(os.tmpdir(), safe_filename);

    fs.writeFileSync(temp_path, pdf_stream.read());

    try {
      const loader = new PDFLoader(temp_path);
      const text_content = await loader.load();

      const pages_content = text_content.map(t => t.pageContent);
      const pages_metadata_content = text_content.map(t => t.metadata);

      const embeddings = await generate_embeddings(pages_content);
      const opensearchClient = new OpenSearchClient(OPENSEARCH_CONFIG);
      await insert_documents(opensearchClient, pages_content, embeddings, pages_metadata_content);
    } finally {
      fs.unlinkSync(temp_path);
    }

    res.json({
      message: 'File uploaded and processed successfully',
      filename: file.originalname,
      size: file.size,
      text_length: text_content.length,
    });
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});
export default router;